mergesort.rb

```diff
@@ -1,27 +1,35 @@
 def mergesort(a)
-    # if the array size is 0|1 then the list is considered sorted, return sorted array
-    # split the list into two halves
-    # merge sort each half
-    # combine the sorted halves
+  # base case
+  return a if a.length <= 1
+  # split array
+  left, right = split_array(a)
+  # recursive case
+  combine(mergesort(left), mergesort(right))
 end
 
 def split_array(a)
-    # find the middle index
-    # split the array in half - take a look at take and drop in the ruby docs, or use ranges
-    # return left and right halves of array as separate arrays - did you know a method can return multiple arrays? look it up
+  middle = a.length / 2
+  left = a[0...middle]
+  right = a[middle..-1]
+  return left, right
 end
 
-# precondition: a and b are sorted
+
 def combine(a, b)
-    # create a results array
-    # counters pointing to the index of the smallest elements in each array
-    # check that we have elements to compare
-        # push the smaller element onto the result array
-    # if there are elements left over in a, move them to result
-    # if there are elements left over in b, move them to result
+  result = []
+  until a.length == 0 || b.length == 0
+    if a.first <= b.first
+      result << a.shift
+    else
+      result << b.shift
+    end
+  end
+  result + a + b
 end
 
```

Comment:
Thank you!

On Tue, May 17, 2016 at 9:15 PM, malderi <notifications@github.com> wrote:

> In mergesort.rb
> <https://github.com/Ada-C5/mergesort/pull/7#discussion_r63642897>:
>
> >  def combine(a, b)
> > -    # create a results array
> > -    # counters pointing to the index of the smallest elements in each array
> > -    # check that we have elements to compare
> > -        # push the smaller element onto the result array
> > -    # if there are elements left over in a, move them to result
> > -    # if there are elements left over in b, move them to result
> > +  result = []
> > +  until a.length == 0 || b.length == 0
> > +    if a.first <= b.first
> > +      result << a.shift
> > +    else
> > +      result << b.shift
> > +    end
> > +  end
> > +  result + a + b
> >  end
> >
>
> Good solution all-around; concise and readable. On line 27, even though
> Ruby does turn the last statement into a return, it is generally more
> readable to other people if you explicitly say "return" (so, "return result
> + a + b"). Not necessarily but tends to be good style.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Ada-C5/mergesort/pull/7/files/37e88d08387fff8a0fb045b473ee72754903d4d0#r63642897>
>